plugins/zynaddsubfx/zynaddsubfx/doc/zynaddsubfx.1.txt
ZYNADDSUBFX(1)
==============
:doctype: manpage

NAME
----
zynaddsubfx - a software synthesizer

SYNOPSIS
--------
*zynaddsubfx* ['OPTIONS']

DESCRIPTION
-----------

*zynaddsubfx* is a polyphonic multimbral synthesizer, which supports three
synthesis engines and numerous effects to generate sound.

1) ADsynth generates sounds by adding a number of voices.
   Each voice has filters, envelopes, LFOs, morphing, modulation
   (Ring Modulation, Phase Modulation... the modulators can have any
   shape), resonance, etc...
   Each voice includes a waveform generator with up to 128 sine/non-sine
   harmonics.
   You can use Fourier synthesis or if you don't like it you
   can wave-shaping/filtering of functions.

2) SUBsynth is a simple engine which makes sounds through
   harmonic filtering of white noise

3) PADsynth is an engine that makes very beautiful pads and other instruments,
   which can be exported for use with other programs.

Midi and audio support exists for OSS, ALSA, JACK, and others...

OPTIONS
-------

*-h, --help*::
    Display command-line help and exit
*-v, --version*::
    Display version and exit
*-l, --load=FILE*::
    Loads a .xmz file
*-L, --load-instrument*=FILE::
    Loads a .xiz file
*-r --sample-rate*=SR::
    Set the sample rate SR
*-b, --buffer-size*=SR::
    Set the buffer size, which determines the granularity of how often parameter
    changes can be applied
*-o, --oscil-size*=OS::
    Set the ADsynth oscillator size
*-S, --swap*::
    Swap Left and Right output channels
*-D, --dump*::
    Dumps midi note ON/OFF commands
*-U, --no-gui*::
    Run ZynAddSubFX without user interface
*-N, --named*=Name::
    Postfix IO Name when possible
*-a, --auto-connect*::
    AutoConnect when using JACK
*-O, --output*=engine::
    Set Output Engine
*-I, --input*=engine::
    Set Input Engine
*-e, --exec-after-init*=command::
    Run post-initialization script. This script will be run after midi and audio
    drivers have been initialized.

BUGS
----
Please report any bugs to either the mailing list
    zynaddsubfx-user@lists.sourceforge.net
or the bugtracker
    http://sourceforge.net/tracker/?group_id=62934

AUTHOR
------
ZynAddSubFX was originally written by Nasca Octavian Paul. It is currently being
maintained by Mark McCurry.


ChangeLog
2009-02-24	Tobias Doerffel <tobydox/at/users/dot/sourceforge/dot/net>

	* plugins/zynaddsubfx/src/Params/PresetsStore.C:
	* plugins/zynaddsubfx/src/Synth/OscilGen.C:
	* plugins/zynaddsubfx/src/UI/OscilGenUI.cc:
	* plugins/zynaddsubfx/src/UI/OscilGenUI.fl:
	follow recent changes in ZynAddSubFX CVS:
	- Fix improper deallocation in PresetsStore
	- Fixed errors with drawing of the Oscillator as reported with valgrind

	* CMakeLists.txt:
	* data/presets/CMakeLists.txt:
	* plugins/CMakeLists.txt:
	* plugins/zynaddsubfx/zynaddsubfx.cpp:
	* plugins/zynaddsubfx/CMakeLists.txt:
	integrated ZynAddSubFX plugin from LMMS-extras into LMMS mainline

2009-02-23	Tobias Doerffel <tobydox/at/users/dot/sourceforge/dot/net>

	* cmake/modules/FindSTK.cmake:
	* CMakeLists.txt:
	reflect changes in MinGW crosscompiling environment, typo fixes

2009-02-21	Paul Giblock <drfaygo/at/gmail/dot/com>

	* include/cusis_style.h:
	* src/gui/cusis_style.cpp:
	Make ScrollBar-hover work again in Qt-4.5

2009-02-20	Paul Giblock <drfaygo/at/gmail/dot/com>

	* include/gui/tracks/pattern_item.h:
	* include/gui/tracks/track_content_object_item.h:
	* include/gui/tracks/track_item.h:
	* include/gui/tracks/track_container_scene.h:
	* include/gui/tracks/bb_tco_item.h:
	* src/gui/classic_style.cpp:
	* src/gui/tracks/track_content_object_item.cpp:
	* src/gui/tracks/track_item.cpp:
	* src/gui/tracks/track_container_scene.cpp:
	* src/gui/tracks/bb_tco_item.cpp:
	* src/gui/tracks/pattern_item.cpp:
	Fix File headers

	* src/gui/cusis_style.cpp:
	Fix cache collision bug 

	* src/gui/tracks/pattern_item.cpp:
	Place marks once-per-tact

	* src/gui/tracks/bb_tco_item.cpp:
	* src/gui/tracks/track_content_object_item.cpp:
	* src/gui/cusis_style.cpp:
	Add per-TCO coloring and non-linear motion when dragging

	* src/gui/classic_style.cpp:
	Fix text clipping

	* gui/lfo_controller_dialog.cpp:
	QT4.5 fixes our error where Controllers couldn't have backgrounds,
	re-add artwork.

2009-02-19	Paul Giblock <drfaygo/at/gmail/dot/com>
	
	* include/track.h:
	* include/pattern.h:
	Fixed constness

	* include/lmms_style.h:
	* include/classic_style.h:
	* include/cusis_style.h:
	* src/gui/tracks/track_content_object_item.cpp:
	* src/gui/tracks/bb_tco_item.cpp:
	* src/gui/tracks/pattern_item.cpp:
	Move TCO drawing into LMMS style

	* include/lcd_spinbox.h:
	* include/combobox.h:
	Update copyright

	* include/gui/tracks/pattern_item.h:
	* include/gui/tracks/track_content_object_item.h:
	* include/gui/tracks/track_item.h:
	* include/gui/tracks/track_container_scene.h:
	* include/gui/tracks/bb_tco_item.h:
	* src/gui/tracks/track_content_object_item.cpp:
	* src/gui/tracks/track_item.cpp:
	* src/gui/tracks/track_container_scene.cpp:
	* src/gui/main_window.cpp:
	Code style and cleanup

	* src/gui/main_window.cpp:
	Update for model move and resize

	* src/gui/classic_style.cpp:
	Add Classic TCO style back, unoptimized at the moment

	* src/gui/cusis_style.cpp:
	Move Shiny-TCO to Cusis and add pixmap caching

2009-02-19	Tobias Doerffel <tobydox/at/users/dot/sourceforge/dot/net>

	* include/fade_button.h:
	* src/gui/widgets/fade_button.cpp:
	rewrote timing of fading animation for not postponing updates of
	hidden fadeButton until it becomes visible

	* src/gui/widgets/automatable_button.cpp:
	* src/core/track.cpp:
	make sure all buttons in trackOperationsWidget have focus-policy
	set to Qt::NoFocus in order to make space play song even if you clicked
	e.g. a mute-button (closes #2486211)

	* include/note.h:
	* src/core/note.cpp:
	- coding style fixes
	- improved handling of shared detuningHelper object

	* include/instrument_track.h:
	* include/piano_roll.h:
	* src/gui/piano_roll.cpp:
	* src/gui/song_editor.cpp:
	* src/tracks/instrument_track.cpp:
	fixed various bugs regarding recording:
	- once recording has been started, stop it as soon as playMode of song
	  changes or similiar things happen (closes #2486341)
	- fixed broken record-accompany in most use-cases other than using
	  record-while-playing-song by also tracking NoteOn events and remember
	  current play position to use the correct note postitions on NoteOff
	  events (play position might have wrapped around in the meantime due
	  to looppoints or end of BB track) (closes #2486299, #2486203)
	- move timeLine marker in Piano Roll and autoscroll when recording while
	  playing song (closes #2486334)

2009-02-14	Andrew Kelley	<superjoe30/at/gmail/dot/com>

	* src/gui/dialogs/about_dialog.ui:	
	updated copyright and added plug to ##lmms per skiessi's request

2009-02-14	Tobias Doerffel <tobydox/at/users/dot/sourceforge/dot/net>

	* src/3rdparty/samplerate/samplerate.h:
	reverted rev 2037 as we have custom changes in this file

2009-02-13	Tobias Doerffel	<tobydox/at/users/dot/sourceforge/dot/net>

	* include/pattern.h:
	* src/tracks/pattern.cpp:
	- added missing slot declaration for abortFreeze(), fixes lockup when
	  freezing pattern (closes #2530320)
	- update patternView after freezing
	- fixed drawing of frozen pattern

	* include/sample_buffer.h:
	* src/core/sample_buffer.cpp:
	- when resampling whole sampleBuffer, set end_of_input = 1 in
	  libsamplerate data struct - fixes lots of zero samples at the end
	  of output buffer (closes #2531452)
	- coding style fixes
	- removed obsolete save/restore of EOF-property in libsample rate
	  data struct when resampling successively (always use 0)

	* src/3rdparty/samplerate/src_linear.c:
	* src/3rdparty/samplerate/samplerate.h:
	* src/3rdparty/samplerate/src_sinc.c:
	* src/3rdparty/samplerate/src_zoh.c:
	* src/3rdparty/samplerate/samplerate.c:
	* CMakeLists.txt:
	integrated latest libsamplerate which is both faster and more reliable

	* src/gui/widgets/visualization_widget.cpp:
	disable output monitor per default and show click-hint

	* include/side_bar.h:
	* include/kmultitabbar.h:
	* include/file_browser.h:
	* include/piano_roll.h:
	* include/track_container_view.h:
	* include/main_window.h:
	* include/fx_mixer_view.h:
	* include/automation_editor.h:

plugins/zynaddsubfx/zynaddsubfx.cpp
/*
 * zynaddsubfx.cpp - ZynAddSubFX-embedding plugin
 *
 * Copyright (c) 2008-2009 Tobias Doerffel <tobydox/at/users.sourceforge.net>
 * 
 * This file is part of Linux MultiMedia Studio - http://lmms.sourceforge.net
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program (see COPYING); if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301 USA.
 *
 */


#include "lmmsconfig.h"

#include <Qt/QtXml>
#include <QtCore/QTemporaryFile>
#include <QtGui/QDropEvent>
#include <QtGui/QPushButton>

#include "zynaddsubfx.h"
#include "engine.h"
#include "mmp.h"
#include "instrument_play_handle.h"
#include "InstrumentTrack.h"
#include "gui_templates.h"
#include "string_pair_drag.h"
#include "remote_zynaddsubfx.h"

#include "embed.cpp"
#include "moc_zynaddsubfx.cxx"


extern "C"
{

Plugin::Descriptor PLUGIN_EXPORT zynaddsubfx_plugin_descriptor =
{
	STRINGIFY( PLUGIN_NAME ),
	"ZynAddSubFX",
	QT_TRANSLATE_NOOP( "pluginBrowser",
			"Embedded ZynAddSubFX" ),
	"Tobias Doerffel <tobydox/at/users.sf.net>",
	0x0100,
	Plugin::Instrument,
	new PluginPixmapLoader( "logo" ),
	"xiz",
	NULL,
} ;

}



zynAddSubFx::zynAddSubFx( InstrumentTrack * _instrumentTrack ) :
	Instrument( _instrumentTrack, &zynaddsubfx_plugin_descriptor ),
	m_plugin( NULL )
{
	initRemotePlugin();

	// now we need a play-handle which cares for calling play()
	InstrumentPlayHandle * iph = new InstrumentPlayHandle( this );
	engine::getMixer()->addPlayHandle( iph );

	connect( engine::getMixer(), SIGNAL( sampleRateChanged() ),
			this, SLOT( updateSampleRate() ) );
}




zynAddSubFx::~zynAddSubFx()
{
	engine::getMixer()->removePlayHandles( instrumentTrack() );

	m_pluginMutex.lock();
	delete m_plugin;
	m_pluginMutex.unlock();
}




void zynAddSubFx::saveSettings( QDomDocument & _doc,
	                             QDomElement & _this )
{
	QTemporaryFile tf;
	if( tf.open() )
	{
		m_plugin->lock();
		m_plugin->sendMessage(
			RemotePlugin::message( IdSaveSettingsToFile ).
				addString(
					QSTR_TO_STDSTR(
						QDir::toNativeSeparators( tf.fileName() ) ) ) );
		m_plugin->waitForMessage( IdSaveSettingsToFile );
		m_plugin->unlock();
		QByteArray a = tf.readAll();
		// remove first blank line
		a.remove( 0,
#ifdef LMMS_BUILD_WIN32
				2
#else
				1
#endif
					);
		QDomDocument doc( "mydoc" );
		doc.setContent( a );
		_this.appendChild( doc.documentElement() );
	}
}




void zynAddSubFx::loadSettings( const QDomElement & _this )
{
	if( !_this.hasChildNodes() )
	{
		return;
	}

	QDomDocument doc;
	doc.appendChild( doc.importNode( _this.firstChild(), true ) );
	QTemporaryFile tf;
	tf.setAutoRemove( false );
	if( tf.open() )
	{
		QByteArray a = doc.toString( 0 ).toUtf8();
		a.prepend( "<?xml version=\"1.0\"?>\n" );
		tf.write( a );
		m_plugin->lock();
		m_plugin->sendMessage(
			RemotePlugin::message( IdLoadSettingsFromFile ).
				addString(
					QSTR_TO_STDSTR(
						QDir::toNativeSeparators( tf.fileName() ) ) ) );
		m_plugin->waitForMessage( IdLoadSettingsFromFile );
		m_plugin->unlock();

		emit settingsChanged();
	}
}




void zynAddSubFx::loadFile( const QString & _file )
{
	m_plugin->lock();
	m_plugin->sendMessage(
		RemotePlugin::message( IdLoadPresetFromFile ).
								addString( QSTR_TO_STDSTR( _file ) ) );
	m_plugin->waitForMessage( IdLoadPresetFromFile );
	m_plugin->unlock();

	emit settingsChanged();
}




QString zynAddSubFx::nodeName() const
{
	return zynaddsubfx_plugin_descriptor.name;
}




void zynAddSubFx::play( sampleFrame * _buf )
{
	m_pluginMutex.lock();
	m_plugin->process( NULL, _buf );
	m_pluginMutex.unlock();
	instrumentTrack()->processAudioBuffer( _buf,
				engine::getMixer()->framesPerPeriod(), NULL );
}




bool zynAddSubFx::handleMidiEvent( const midiEvent & _me,
                                                const midiTime & _time )
{
	m_pluginMutex.lock();
	m_plugin->processMidiEvent( _me, 0 );
	m_pluginMutex.unlock();

	return true;
}




void zynAddSubFx::updateSampleRate()
{
	m_pluginMutex.lock();

	multimediaProject m( multimediaProject::InstrumentTrackSettings );
	saveSettings( m, m.content() );

	initRemotePlugin();

	loadSettings( m.content() );

	m_pluginMutex.unlock();
}




void zynAddSubFx::initRemotePlugin()
{
	delete m_plugin;
	m_plugin = new RemotePlugin( "remote_zynaddsubfx", false );
	m_plugin->lock();
	m_plugin->waitForInitDone( false );

	m_plugin->sendMessage(
		RemotePlugin::message( IdZasfPresetDirectory ).
			addString(
				QSTR_TO_STDSTR(
					QString( configManager::inst()->factoryPresetsDir() +
								QDir::separator() + "ZynAddSubFX" ) ) ) );
	m_plugin->unlock();
}




PluginView * zynAddSubFx::instantiateView( QWidget * _parent )
{
	return new ZynAddSubFxView( this, _parent );
}







ZynAddSubFxView::ZynAddSubFxView( Instrument * _instrument, QWidget * _parent ) :
	InstrumentView( _instrument, _parent )
{
	setAutoFillBackground( true );
	QPalette pal;
	pal.setBrush( backgroundRole(), PLUGIN_NAME::getIconPixmap(
								"artwork" ) );
	setPalette( pal );

	m_toggleUIButton = new QPushButton( tr( "Show GUI" ), this );
	m_toggleUIButton->setCheckable( true );
	m_toggleUIButton->setChecked( false );
	m_toggleUIButton->setGeometry( 45, 80, 160, 24 );
	m_toggleUIButton->setIcon( embed::getIconPixmap( "zoom" ) );
	m_toggleUIButton->setFont( pointSize<8>( m_toggleUIButton->font() ) );
	connect( m_toggleUIButton, SIGNAL( toggled( bool ) ), this,
							SLOT( toggleUI() ) );
	m_toggleUIButton->setWhatsThis(
		tr( "Click here to show or hide the graphical user interface "
			"(GUI) of ZynAddSubFX." ) );

	setAcceptDrops( true );
}




ZynAddSubFxView::~ZynAddSubFxView()
{
}




void ZynAddSubFxView::dragEnterEvent( QDragEnterEvent * _dee )
{
	if( _dee->mimeData()->hasFormat( stringPairDrag::mimeType() ) )
	{
		QString txt = _dee->mimeData()->data(
						stringPairDrag::mimeType() );
		if( txt.section( ':', 0, 0 ) == "pluginpresetfile" )
		{
			_dee->acceptProposedAction();
		}
		else
		{
			_dee->ignore();
		}
	}
	else
	{
		_dee->ignore();
	}
}




void ZynAddSubFxView::dropEvent( QDropEvent * _de )
{
	const QString type = stringPairDrag::decodeKey( _de );
	const QString value = stringPairDrag::decodeValue( _de );
	if( type == "pluginpresetfile" )
	{
		castModel<zynAddSubFx>()->loadFile( value );
		_de->accept();
		return;
	}
	_de->ignore();
}




void ZynAddSubFxView::modelChanged()
{
	toggleUI();
}




void ZynAddSubFxView::toggleUI()
{
	if( m_toggleUIButton->isChecked() )
	{
		castModel<zynAddSubFx>()->m_plugin->showUI();
	}
	else
	{
		castModel<zynAddSubFx>()->m_plugin->hideUI();
	}

}





extern "C"
{

// neccessary for getting instance out of shared lib
Plugin * PLUGIN_EXPORT lmms_plugin_main( Model *, void * _data )
{

	return new zynAddSubFx( static_cast<InstrumentTrack *>( _data ) );
}


}


